<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Alpha Diversity</title>

<script src="site_libs/header-attrs-2.27/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Microbiome Data Analysis</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="about.html">Home</a>
</li>
<li>
  <a href="pkgs.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Descriptives
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="prepro.html">Preprocessing</a>
    </li>
    <li>
      <a href="alphadiv.html">alpha diversity</a>
    </li>
    <li>
      <a href="betadiv.html">beta diversity</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Statistical inference
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="DA.html">Differential Abundance</a>
    </li>
    <li>
      <a href="DAmixed.html">Differential Abundance - Repeated measures</a>
    </li>
    <li>
      <a href="adonistest.html">ANOVA on beta diversity</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Supervised Multivariate models
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="hm.html">Heatmaps</a>
    </li>
    <li>
      <a href="cca.html">Canonical Correlation Analysis</a>
    </li>
    <li>
      <a href="mediationanalysis.html">Mediation analysis via Propensity Scoring</a>
    </li>
  </ul>
</li>
<li>
  <a href="https://github.com/mortenarendt/MicrobiomeDataAnalysis/tree/master/data">Data</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Alpha Diversity</h1>

</div>


<style>
pre.bluecars {
    background-color: #aabbff !important;
}
pre.blackcars {
    background-color: #000000 !important;
}
pre.redcars {
    background-color: #ffbbbb !important;
}
</style>
<p>For each sample a single measure can be calculated describing the
diversity within the sample. The most intuitive is <em>how many
different bacteria (OTUs) is observed</em>. But there are others. The
phyloseq package with the function plot_richness() supports seven
different versions.</p>
<p>These can easily be calculated and visualized according to a given
design.</p>
<pre class="r"><code>library(ggplot2)
library(ggpubr)
library(phyloseq)
load(&#39;./data/Mice_csec.RData&#39;)

pp &lt;- plot_richness(phyX)
AlphaD &lt;- pp$data

ggplot(data = AlphaD, aes(Birth_mode, value, fill = Birth_mode)) + 
  geom_boxplot() + 
  geom_jitter() + 
  stat_compare_means(vjust = 1,method = &#39;t.test&#39;) + 
  facet_wrap(~variable, nrow = 2, scales = &#39;free&#39;) + 
  theme(legend.position = &#39;top&#39;)</code></pre>
<p><img src="alphadiv_files/figure-html/unnamed-chunk-1-1.png" width="672" /></p>
<p>What you see here, is some indications of <em>vaginal delivery</em>
results in lower diversity. However, due to the low n significance is
hard to achieve.</p>
<p>Shannon diversity is a metric with relatively high dependency on rare
OTUs. <em>Simpson</em> diversity (and <em>Inverse Simpson</em>) depends
on the even-ness of the observed OTUs in the samples and is hence
relatively independent on rare species, as is obvious from the
mathematical formula. All three measures seems to be related to birth
mode.</p>
<div id="faiths-phylogenetic-diversity"
class="section level2 unnumbered">
<h2 class="unnumbered">Faiths Phylogenetic Diversity</h2>
<p>The alpha diversity metrics computed in phyloseq is <em>not</em>
taking phylogentics into account. However, there is one commenly used
method <em>Faiths PD</em> which can be calculated using the CRAN package
picante. The calculation is easy, but adding it to all the other
metrics, is a bit tricky:</p>
<pre class="r"><code>library(picante)
PD &lt;- pd(t(phyX@otu_table),phyX@phy_tree, include.root = F)
## Add it to the sample data 
PD &lt;- cbind(sample_data(phyX) %&gt;% data.frame() %&gt;% rownames_to_column(&#39;samples&#39;),PD) %&gt;% 
  gather(variable,value,PD,SR) 
# ... and merge with the rest of alpha diversity values
AlphaD &lt;- AlphaD %&gt;% bind_rows(PD)</code></pre>
<p>In order to get more hands-on experience with alpha diversity and
especially how different preprocessing techniques influences these
metrics, the exercise below show case how inherent the stocastisity of
the sequencing depth is in the data.</p>
</div>
<div id="exercise" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Exercise</h1>
<div id="with-no-prepro" class="section level2" number="1.1">
<h2><span class="header-section-number">1.1</span> With No prepro</h2>
<p>Rerun the above alpha diversity analysis with no preprocessing of the
data</p>
</div>
<div id="alpha-vs-library-size" class="section level2" number="1.2">
<h2><span class="header-section-number">1.2</span> Alpha vs library
size</h2>
<p>Library size (number of reads per sample), is in amplicon sequencing
considered to be nuisance with NO biological interpretation at all.
Compare each measure of alpha diversity with the library size. What is
problematic? (You can take inspiration in the code below on how to do
this in R)</p>
<pre class="r"><code>libsize &lt;- data.frame(libsize = sample_sums(phyX))

# merge
AlphaD$X.SampleID &lt;- as.character(trimws(AlphaD$X.SampleID))
AlphaD &lt;- merge(AlphaD, libsize, by.x = &#39;X.SampleID&#39;, by.y = &#39;row.names&#39;)

ggplot(data = AlphaD, aes(libsize,value)) + 
  geom_point() + 
  stat_smooth(method = lm) + 
  facet_wrap(~variable, scales = &#39;free&#39;)</code></pre>
<p>Two preprocessing techniques are investigated in terms their ability
to remove the dependency on the library size. Namely rarefaction and
OTU-filtering.</p>
</div>
<div id="rarefy" class="section level2" number="1.3">
<h2><span class="header-section-number">1.3</span> Rarefy</h2>
<p>Rarefy the sample to even depth, and calculate the corresponding
alpha diversity measures</p>
<pre class="r"><code>phyXrare &lt;- rarefy_even_depth(phyX)
AlphaDrare &lt;- plot_richness(phyXrare)$data</code></pre>
</div>
<div id="does-it-work" class="section level2" number="1.4">
<h2><span class="header-section-number">1.4</span> Does it work?</h2>
<p>Merge with the <strong>original library size</strong> and check if
you had succes.</p>
<pre class="r"><code>AlphaDrare$X.SampleID &lt;- as.character(trimws(AlphaDrare$X.SampleID))
AlphaDrare &lt;- merge(AlphaDrare, libsize, by.x = &#39;X.SampleID&#39;, by.y = &#39;row.names&#39;)

ggplot(data = AlphaDrare, aes(....)) + </code></pre>
</div>
<div id="filter-of-rare-taxa" class="section level2" number="1.5">
<h2><span class="header-section-number">1.5</span> Filter of rare
taxa</h2>
<p>Now try to use a filter that removes rare taxa</p>
<pre class="r"><code>phyXfilt &lt;- filter_taxa(phyX,function(x) sum(x&gt;0) &gt; ??? , TRUE)

AlphaDfilt &lt;- plot_richness(phyXfilt)$data
AlphaDfilt$X.SampleID &lt;- as.character(trimws(AlphaDfilt$X.SampleID))
AlphaDfilt &lt;- merge(AlphaDfilt, libsize, by.x = &#39;X.SampleID&#39;, by.y = &#39;row.names&#39;)
ggplot(data = AlphaDfilt, aes(...)) + </code></pre>
<p>What you have learned here is that sequencing depth - and its
variation between samples - inherently affects data analysis.
Preprocessing is capable of handling these issues, but <strong>only to
some extend</strong>.</p>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="unnumbered">References</h1>
<p>Xia, Yinglin, Jun Sun, and Ding-Geng Chen. <strong>Statistical
analysis of microbiome data with R</strong>. Springer, 2018. Chapter
6.</p>
<p>Faith D.P. (1992) <strong>Conservation evaluation and phylogenetic
diversity</strong>. Biological Conservation, 61, 1-10.</p>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
