<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Canonical Correlation Analysis</title>

<script src="site_libs/header-attrs-2.24/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Microbiome Data Analysis</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="about.html">Home</a>
</li>
<li>
  <a href="pkgs.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Descriptives
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="prepro.html">Preprocessing</a>
    </li>
    <li>
      <a href="alphadiv.html">alpha diversity</a>
    </li>
    <li>
      <a href="betadiv.html">beta diversity</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Statistical inference
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="DA.html">Differential Abundance</a>
    </li>
    <li>
      <a href="DAmixed.html">Differential Abundance - Repeated measures</a>
    </li>
    <li>
      <a href="adonistest.html">ANOVA on beta diversity</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Supervised Multivariate models
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="hm.html">Heatmaps</a>
    </li>
    <li>
      <a href="cca.html">Canonical Correlation Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="https://github.com/mortenarendt/MicrobiomeDataAnalysis/tree/master/data">Data</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Canonical Correlation Analysis</h1>

</div>


<p>In order to answer biological questions, often a combination of
high-throughput data is generated. Especially in combination with
microbiome types of data, the associated metabolome is naturally of
interest, as these two sources together reflects <em>who are there</em>
and <em>what do they do</em>.</p>
<p>In the analysis of such data a natural starting point is to look for
common structure. That is; which types of bacteria are correlated to a
certain type of metabolites.</p>
<div id="canonical-correlation-analysis"
class="section level1 unnumbered">
<h1 class="unnumbered">Canonical Correlation Analysis</h1>
<p>Canonical correlation analysis is a rather old statistical technique
for finding common components between two sets of correlated variables
such as microbiome and metabolomics data.</p>
<p>For an introduction and some <em>getting started</em> code check out
the <a href="http://mixomics.org/methods/rcca/">mixOmics package
webpage</a>.</p>
<p>Indeed the vanilla CCA is limited due to dimentionallity issues from
way to many variables compared to number of samples. Therefore several
attemps has been made to circumvent this. Here we will exchange
correlation by covariance to add a simple fix.</p>
<p>Below is a 2 component canonical covariance model.</p>
<p>But first you need to install a package.</p>
<pre class="r"><code>library(devtools)
install_github(&#39;mortenarendt/StructuralKnowledgeModl&#39;)</code></pre>
<pre class="r"><code>library(structMultMdl)
library(phyloseq)
library(tidyverse)
load(&#39;./data/Rats_inulin.RData&#39;)

phyXpp &lt;- subset_samples(phyX, !is.na(Acetate))
phyXpp &lt;- filter_taxa(phyXpp, function(x) sum(x&gt;0)&gt;0, TRUE)
phyXpp &lt;- transform_sample_counts(phyXpp, function(x) x/sum(x))

OTUtb &lt;- data.frame(t(otu_table(phyXpp)))

Mtb &lt;- sample_data(phyXpp)[,c(&quot;Acetate&quot;,&quot;Butyrate&quot;,&quot;Fructose&quot;,&quot;Glucose&quot;,&quot;Lactate&quot;,&quot;Propionate&quot;,&quot;Xylose&quot;)]
# preprocess the Mtb data
Mtb$Xylose &lt;- Mtb$Xylose + 0.141/2 # there is zeros for this one - so to fix log we need to add a bit. 
Mtb &lt;- log(Mtb)

# scale the variables
GM &lt;- as.matrix(scale(OTUtb, center = T))
mtb &lt;- as.matrix(scale(Mtb, center = T))

results &lt;- pCCA(mtb,GM,ncomp = 2, na = 0,nperm = 3)

print(results$res_df)</code></pre>
<pre><code>##   A p_corr_cmp1 p_corr_cmp2 p_var_cmp1 p_var_cmp2
## 1 0           0           0          0          0</code></pre>
<p>Now, lets try to extract the model parameters and plot them</p>
<pre class="r"><code># dig out scores and loadings and plotit. 
scores &lt;- data.frame(tx = results$a_results[[1]]$model$tx,
                     ty = results$a_results[[1]]$model$ty)

loadsY &lt;- data.frame(wy = results$a_results[[1]]$model$Wy)
loadsY &lt;- cbind(loadsY, tax_table(phyXpp), by = &#39;row.names&#39;)

loadsX &lt;- data.frame(wx = results$a_results[[1]]$model$Wx) 
loadsX$name &lt;- rownames(loadsX)

ggplot(data = scores, aes(tx.1,ty.1)) + 
  geom_point() + 
  xlab(&#39;Metabolomics Score&#39;) + ylab(&#39;Microbiome Score&#39;) + 
  stat_smooth(method = lm, se= F)</code></pre>
<p><img src="cca_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
<pre class="r"><code>ggplot(data = loadsY[order(loadsY$Rank4),],aes(x = 1:476,wy.1,fill = Rank3)) + 
  geom_bar(stat = &#39;identity&#39;) + 
  ggtitle(&#39;Microbiome loadings&#39;) + 
  theme(legend.position = &#39;bottom&#39;)</code></pre>
<p><img src="cca_files/figure-html/unnamed-chunk-3-2.png" width="672" /></p>
<pre class="r"><code>ggplot(data = loadsX,aes(wx.1,wx.2,label = name)) + 
  geom_point() + geom_label() + 
  ggtitle(&#39;Metabolomics loadings&#39;) + 
  geom_hline(yintercept = 0) + geom_vline(xintercept = 0)</code></pre>
<p><img src="cca_files/figure-html/unnamed-chunk-3-3.png" width="672" /></p>
<p>This is a nice correlation, and this model is probably a good model,
slightly overfitted though, but the main concern here is that it is hard
to interpret, as <em>ALL</em> the OTUs are active and further, tends to
be pointing in different direction eventhough they are rather similar
wrt taxonomy.</p>
</div>
<div id="penalized-canonical-correlation-analysis"
class="section level1 unnumbered">
<h1 class="unnumbered">Penalized Canonical Correlation Analysis</h1>
<p>In order to circumvent this, a penalized model is produced. Two types
of penalties are incorporated;</p>
<ul>
<li>Sparsity - to set a fair part of the <em>not so informative</em>
loadings to zero.</li>
<li>Loading similarity for phylogentic similarity.</li>
</ul>
<p>The heuristics behind a phylogentic penalty is to make OTUs which are
phylogentically close also to have similar model parameters. Here, a the
phylogenetic tree is represented as a similarity matrix, which can be
extracted by the cophenetic.phylo() function from the library ape, and
translated from distance to similarity by either cor() or
exp(-distance).</p>
<p>The sumabs argument is controling the level of sparsity for each
loading vector on the input matrices respectively, such that a low value
gives a more sparse model. the <strong>na</strong> argument imposes a
grid from 0 (no phylogentic penalty) to 1 (alot of phylogentic penalty),
resulting in just as many models.</p>
<pre class="r"><code>fixKernelMatrix &lt;- function(P, thr = 1e-5,niter = 1000){
  conv &lt;- 1
  n &lt;- dim(P)[1]
  cc &lt;- 0
  PP &lt;- P
  while (conv&gt;thr &amp; cc&lt;niter){
    cc &lt;- cc+1
    for (i in 1:n){
      PP[i,] &lt;- P[i,] / sum(P[i,])
      PP[,i] &lt;- P[,i] / sum(P[,i])
    }
    
    conv &lt;- norm(P - PP,&#39;F&#39;)
    P &lt;- PP
  }
  print(paste(&#39;Kernel Smoother updated in&#39;,cc, &#39;iterations at thr =&#39;,thr))
  return(PP)
}


D &lt;- ape::cophenetic.phylo(phy_tree(phyXpp))
#R &lt;- cor(D)
R &lt;- fixKernelMatrix(exp(-D))</code></pre>
<pre><code>## [1] &quot;Kernel Smoother updated in 220 iterations at thr = 1e-05&quot;</code></pre>
<pre class="r"><code>res_pen &lt;- pCCA(mtb,GM,R,ncomp = 2,sumabs = c(10,6), na = 10,nperm = 30)

print(res_pen$res_df)</code></pre>
<pre><code>##            A p_corr_cmp1 p_corr_cmp2 p_var_cmp1 p_var_cmp2
## 1  0.0000000           0  0.00000000          0 0.03333333
## 2  0.1111111           0  0.00000000          0 0.00000000
## 3  0.2222222           0  0.03333333          0 0.00000000
## 4  0.3333333           0  0.00000000          0 0.00000000
## 5  0.4444444           0  0.00000000          0 0.00000000
## 6  0.5555556           0  0.06666667          0 0.00000000
## 7  0.6666667           0  0.00000000          0 0.00000000
## 8  0.7777778           0  0.00000000          0 0.03333333
## 9  0.8888889           0  0.00000000          0 0.00000000
## 10 1.0000000           0  0.16666667          0 0.36666667</code></pre>
<p>Now we take a descision on which model to use and extract it in a
similar fashion as above. Resulting in the microbiome loadingplot
below.</p>
</div>
<div id="exercises" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Exercises</h1>
<div id="vanilla-cca" class="section level2" number="1.1">
<h2><span class="header-section-number">1.1</span> Vanilla CCA</h2>
<p>Perform the <strong>uncontrained</strong> CCA as above and
reconstruct the plots</p>
</div>
<div id="penalized-cca" class="section level2" number="1.2">
<h2><span class="header-section-number">1.2</span> Penalized CCA</h2>
<p>Try build a penalized version with two components, no sparsity for
the metabolites but some sparsity for the microbiome. Screen a grid of
phylo-penalities and decide which model to use, extract and plot it.</p>
</div>
</div>
<div id="references" class="section level1" number="2">
<h1><span class="header-section-number">2</span> References</h1>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
